### DISCUSSION TITLE
How to protect cluster from accidental deletion of resources?

### DISCUSSION THREAD

**Discussion by mloskot** [fluxcd/flux2#3743] [General] [answered]
Hi, 

I'm a Kubernetes and Flux newbie learning the cloud-native technologies, and learning it the hard-way apparently😊 

Over last two weeks, I've lost my GitOps/Flux-operated cluster twice by accident.
I'm not entirely clear what happened, but I think I deleted Flux deployment all together, loosing pods Flux controllers pods.
I had eventually been observing cascade effect of all my resources being deleted leaving just PV-s which have reclaim policy set to retain.

I'm trying to understand what did I do wrong exactly and how to protect myself against similar mistakes in future.

I found some related discussions in Flux v1 archives, like https://github.com/fluxcd/flux/issues/3148, and this thread https://github.com/fluxcd/flux2/discussions/2863 as well as this very informative story of [What we learned from InfluxDB](https://www.influxdata.com/blog/deleting-production-steps-how-fix-it/), but I guess there must be others, and even canonical resources I should learn from.

Could anyone point me at the must-read resources on Flux and protecting Kubernetes cluster resources against accidental losses due to misuses or misconfiguration of Flux and GitOps automation?

**Comment by kingdonb** [comment, marked as answer]
Have you figured out what you're doing that causes the deletion?

There is a directive you can apply to prevent resources from being deleted, but if there's something deleted those resources more than once, you definitely have a misunderstanding in your configuration and that should be resolved before you try to paper over it with an annotation.

The `Kustomization` has `spec.prune` which typically should be enabled, so that deleting a resource from the git repo will actually clean up after the deletion and delete the matching resource from the cluster.

Do you know what happened that causes your resource to get deleted?

If there are some high-value resources that you want to ensure never get deleted, there's an annotation for that:

> You can disable pruning for certain resources by either labelling or annotating them with:

```
kustomize.toolkit.fluxcd.io/prune: disabled
```

It's covered here: https://fluxcd.io/flux/components/kustomize/kustomization/#prune

The most fundamental doc is the getting started guide, and the two guides that are linked at the bottom of it. If you have diverged much from what those guides prescribe, then it will be really helpful to see what else you have done, in order to understand why some resources may get deleted inappropriately from time to time.

https://fluxcd.io/flux/get-started/
with the following guides:

https://github.com/fluxcd/flux2-kustomize-helm-example
https://github.com/fluxcd/flux2-multi-tenancy

If you can't provide the repo itself, maybe you can replicate the structure in a public setting where we can have a look at what's going wrong? It has to be some misunderstanding, although it is possible you found a bug, I find this highly unlikely shooting from the hip, it seems a lot more likely to be some overlap due to error or fundamental misconfiguration that makes the error too easy to commit.

There has been talk about more advanced `flux diff` or making it embedded into the PR workflow, so you can read the diff as an attachment to the PR. This might be really helpful in certain configurations, especially where a mismatch in overlay or patch structure can create a fault that prevents resources from being created.

I would 100% not take one step further until you ascertained why the resources got deleted, especially if it's happening more than once.

**Reply by mloskot** [reply]
> Have you figured out what you're doing that causes the deletion?

I have tried to reproduce the sequence on local cluster.
I'm close to being confident that what led to the deletion of almost everything was deletion of the main Flux Kustomization.

**How the deletion of the Kustomization happened?**

First time, I was testing how to apply Kustomize patches to some of `gotk-components.yaml` and observing no patching effects, I intentionally deleted the Kustomization. Lack of knowledge, understanding and experience. Check!

Second time was clearly accidental. I went to Lens > Events with aim to delete some event, but instead of clicking on the `Message` column of the event, I clicked on the "Involved Object" and on the page that opened, I hit the bin button. If I checked the title of the page, I would have noticed the object to be deleted is not `Event` but `Kustomization`. A complete brain failure or too much trust in ClickOps - lesson learned!

> If there are some high-value resources that you want to ensure never get deleted, there's an annotation for that

Yes, I've been learning about it and running some tests.

Currently, I applied the `prune: false` (https://github.com/fluxcd/flux2/discussions/2102#discussioncomment-5542724) and observing how this affects my cluster and my workflow in action.
Next, I am planning to fine-tune the pruning with the annotations. I will want to protect StatefulSet-s at least,
and Flux deployments too :)

> If you can't provide the repo itself, maybe you can replicate the structure in a public setting where we can have a look at what's going wrong?

I can't provide the repo, but next time I hit 

> I would 100% not take one step further until you ascertained why the resources got deleted, especially if it's happening more than once.

This is a golden recommendation indeed. That is why I have started using local clusters with attempt to reproduce my mistakes.

> There has been talk about more advanced flux diff or making it embedded into the PR workflow, so you can read the diff as an attachment to the PR.

Yes, I have learned about `flux diff`, https://github.com/fluxcd/flux2/discussions/1617#discussioncomment-5515362, and I will be looking into utilising its powers.

I have already embraced the power of `validate.sh` from https://github.com/fluxcd/flux2-kustomize-helm-example

Thank you @kingdonb for your notes. Much appreciated!
